Convert notification settings to use new settings_ui.do_settings_change system.

shubhamdhama opened the pull request:
**settings: Fix real-time sync with notifications settings.**
The if condition in "update_global_notifications" case was always false.
Similar to c672a1c.

**settings: Refactor notification settings to use do_settings_change.**
Use settings_ui.do_settings_change system for making all requests
and to show our new settings change status indicator.

Fixes: #8587.

shubhamdhama commented:
Breaking the `notification_settings` array into more arrays makes and use `_.each` for each setting was the better solution I have another option was to use more if statements which don't LGTM. Is there any better way.
I've also fixed the real-time sync.

Manual Testing: I've first `checked` some notifications settings and then in other windows, they were syncing properly and also I've refreshed them to see if they are really changing or not.

GIF: 
![saving changes](https://user-images.githubusercontent.com/22238472/37091040-b8d35996-222c-11e8-9475-cc1dbc6e65ee.gif)
![realtimesync](https://user-images.githubusercontent.com/22238472/37091042-b914cc14-222c-11e8-8c0a-b71a5dd539ae.gif)

timabbott commented:
@shubhamdhama this looks great!  A few notes though before we can merge:
* The second commit should be split into a couple commits for the fundamentally separate changes you're making here: E.g. splitting `notification_settings` should be a separate commit from the UI changes.  I would expect the second commit to split into 2-3 commits.
* Agreed those yes/no buttons should be a different style; I'd recommend adding a new commit at the end to address that.
* For temporarily disabling notifications, I think we want a separate "do not disturb mode" type feature elsewhere in the app.  There's a separate issue for this; it would be a totally reasonable thing to work on soon.

shubhamdhama commented:
@timabbott I've split the second commit into two. The second commit was too dependent inside but still, I've split them into two commits but I wasn't sure about the commit message of the first one.
(I should improve my small commit making skills :/

timabbott commented:
@shubhamdhama there's a better way to do that split, without copying large blocks of code.  Let me show you.

timabbott commented on static/js/settings_notifications.js:
```diff
@@ -2,17 +2,23 @@ var settings_notifications = (function () {
-var notification_settings = [
+var stream_notification_settings = [
+    {setting: "enable_stream_desktop_notifications", notifications:"desktop_notifications"},
+    {setting: "enable_stream_push_notifications", notifications:"desktop_notifications"},
```
This should have been `push_notifications`.  Be very careful when refactoring -- this was a fairly serious bug that would likely not have been caught aside from reading the code. 

(This sort of bug is much more visible with smaller commits, which is why we require them in this project).

timabbott commented:
I fixed the bug noted above, split the commits how I would do it, and merged as the 3 commits ending with 0d29864c6ad9fec4d0f82c12d2331e5bfc4a9a4a.  You should study these to see how to do this well in the future; producing a commit list like this that's readable is a really important skill that really improves your ability to communicate through your code changes in a way that makes bugs shallow and the code easy to merge.  

I'd recommend reviewing our guidelines for more advice on the topic: https://zulip.readthedocs.io/en/latest/contributing/version-control.html

timabbott commented:
Not a big deal; I find I end up needing to teach these skills to most contributors who don't have extensive prior experience in open source, and as you can see from the timestamps, it didn't take me that long :)

And thanks for taking care of this cleanup quickly :).  I think the next step here is we should chat with @rishig (maybe in a new thread on #design) about our strategy for making other elements use a similar feedback system.

Also, I guess you can work on better buttons :)
